damnit/backend/extraction_control.py

```diff
@@ -0,0 +1,247 @@
+"""Machinery to run Slurm jobs for extracting data
+
+See extract_data.py for what happens inside the jobs this launches.
+"""
+import getpass
+import logging
+import os
+import shlex
+import subprocess
+import sys
+from contextlib import contextmanager
+from ctypes import CDLL
+from dataclasses import dataclass
+from pathlib import Path
+from threading import Thread
+
+from extra_data.read_machinery import find_proposal
+
+from .db import DamnitDB
+from ..context import RunData
+
+log = logging.getLogger(__name__)
+
+
+# Python innetgr wrapper after https://github.com/wcooley/netgroup-python/
+def innetgr(netgroup: bytes, host=None, user=None, domain=None):
+    libc = CDLL("libc.so.6")
+    return bool(libc.innetgr(netgroup, host, user, domain))
+
+
+def default_slurm_partition():
+    username = getpass.getuser().encode()
+    if innetgr(b'exfel-wgs-users', user=username):
+        return 'exfel'
+    elif innetgr(b'upex-users', user=username):
+        return 'upex'
+    return 'all'
+
+
+def process_log_path(run, proposal, ctx_dir=Path('.'), create=True):
+    p = ctx_dir.absolute() / 'process_logs' / f"r{run}-p{proposal}.out"
+    if create:
+        p.parent.mkdir(exist_ok=True)
+        if p.parent.stat().st_uid == os.getuid():
+            p.parent.chmod(0o777)
+        p.touch(exist_ok=True)
+        if p.stat().st_uid == os.getuid():
+            p.chmod(0o666)
+    return p
+
+
+@contextmanager
+def tee(path: Path):
+    with path.open('ab') as fout:
+        r, w = os.pipe()
+        def loop():
+            while b := os.read(r, 4096):
+                fout.write(b)
+                sys.stdout.buffer.write(b)
+                sys.stdout.flush()
+
+        thread = Thread(target=loop)
+        thread.start()
+        try:
+            yield w
+        finally:
+            os.close(w)
+            thread.join()
+            os.close(r)
+
+
+def proposal_runs(proposal):
+    proposal_name = f"p{int(proposal):06d}"
+    raw_dir = Path(find_proposal(proposal_name)) / "raw"
+    return set(int(p.stem[1:]) for p in raw_dir.glob("*"))
```

Comment:
This function is already in use, just moved, and so far we haven't had any problem with it. It might make sense to be a bit more cautious here, but I'd prefer to tackle that in a separate PR.